tools/test/framework/framework.go

```diff
@@ -0,0 +1,213 @@
+package framework
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strconv"
+	"testing"
+	"time"
+
+	http_helper "github.com/gruntwork-io/terratest/modules/http-helper"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	"github.com/prometheus/common/log"
+	"github.com/stretchr/testify/require"
+
+	"github.com/gruntwork-io/terratest/modules/k8s"
+	. "github.com/onsi/gomega"
+)
+
+type TestFramework struct {
+	testing.T
+	k8sclusters []string
+	kumactl     string
+	verbose     bool
+}
+
+// NewK8sTest gets the number of the clusters to use in the tests, and the pattern
+// to locate the KUBECONFIG for them. The second argument can be empty
+func NewK8sTest(numClusters int, kubeConfigPathPattern string, verbose bool) *TestFramework {
+	if numClusters < 1 || numClusters > maxClusters {
+		log.Error("Invalid cluster number. Should be in the range [1,3], but it is ", numClusters)
+		return nil
+	}
+	if kubeConfigPathPattern == "" {
+		kubeConfigPathPattern = defaultKubeConfigPathPattern
+	}
+	kubeconfigs := []string{""} // have an empty cluster at [0]
+	for i := 1; i <= numClusters; i++ {
+		kubeconfigs = append(kubeconfigs,
+			os.ExpandEnv(fmt.Sprintf(kubeConfigPathPattern, i)))
+	}
+
+	kumactl := os.Getenv(envKUMACTL)
+	if kumactl == "" {
+		log.Error("Unable to find kumactl, please supply valid KUMACTL environment variable.")
+		return nil
+	}
+
+	return &TestFramework{
+		k8sclusters: kubeconfigs,
+		kumactl:     kumactl,
+		verbose:     verbose,
+	}
+}
+
+func (t *TestFramework) ApplyOnK8sCluster(idx int, namespace string, yaml string) error {
+	options := k8s.NewKubectlOptions("", "", namespace)
+	defer k8s.KubectlDelete(t, options, t.k8sclusters[idx])
+	k8s.KubectlApply(t, options, yaml)
+	return nil
+}
+
+func (t *TestFramework) ApplyAndWaitServiceOnK8sCluster(idx int, namespace string, yaml string, service string) error {
+	options := k8s.NewKubectlOptions("", "", namespace)
+	defer k8s.KubectlDelete(t, options, t.k8sclusters[idx])
+	k8s.KubectlApply(t, options, yaml)
+	k8s.WaitUntilServiceAvailable(t, options, service, defaultRetries, defaultTiemout)
+	return nil
+}
+
+func (t *TestFramework) DeployKumaOnK8sCluster(idx int) {
+	require.NoError(t, t.DeployKumaOnK8sClusterE(idx))
+}
+
+func (t *TestFramework) DeployKumaOnK8sClusterE(idx int) error {
```

Comment:
Can we leave just one version of this method with error?

Reply:
Done, thanks!